### PR TITLE
feat: check restriction on main actor before type checking

### PR DESCRIPTION
**Pull request by crusso**
Improve the error message for compiled code that contains illegal top-level declarations in the main program.

c.f. https://forum.dfinity.org/t/misplaced-await-and-misplaced-async/34479

### CODE REVIEW

File: test/run-drun/bad-actor.mo [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+type T1 = {};
+type T2 = {};
+
+actor Self {
+
+};
+
+// bad trailing decs
+type U1 = {};
+type U2 = {};
```
Comment (crusso) [inline test/run-drun/bad-actor.mo:12]:
```suggestion
type U2 = {};

```

File: test/run-drun/ok/bad-actor.comp-ref.ok [hunk 2]
```diff
@@ -0,0 +1,4 @@
+bad-actor.mo:3.1-4.13: type error [M0141], move these declarations into the body of the main actor or actor class
+bad-actor.mo:11.1-12.13: type error [M0141], move these declarations into the body of the main actor or actor class
+bad-actor.mo:6.1-8.2: type error [M0141], an actor or actor class must be the only non-imported declaration in a program
+  (This is a limitation of the current version and flag -ref-system-api.)
```
Comment (luc-blaeser) [inline test/run-drun/ok/bad-actor.comp-ref.ok:4]: Just a detail, not worth adjusting, maybe something for backlog: Is it really a limitation or maybe it is the clean way of how we encourage it for the future?

**Review by luc-blaeser — APPROVED** [review]
Very nice! Thanks for implementing.

File: test/run-drun/bad-actor.mo [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
+// bad leading decs
+type T1 = {};
+type T2 = {};
+
+actor Self {
```
Comment (ggreif) [inline test/run-drun/bad-actor.mo:6]: Perhaps move to `test/fail`? Or could we later define a semantics to consider such outside declarations to be in the actor scope? Many people tend to think this to be natural style.

**Review by ggreif — APPROVED** [review]
Very nice rationalisation of error messages!

File: test/run-drun/ok/bad-actor.comp-ref.ok (hunk 2 above)
Comment (crusso) [reply test/run-drun/ok/bad-actor.comp-ref.ok:4]: Yeah, it seems weird to allow declaration of state outside the actor - maybe we could relax in future and only allow static declarations in the prelude.

**Pull request merged**